// @aws-cdk/aws-cloudwatch-actions/lib/ec2.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ec2Action = exports.Ec2InstanceAction = void 0;
const jsiiDeprecationWarnings = require("../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("@aws-cdk/core");
/**
 * Types of EC2 actions available
 */
var Ec2InstanceAction;
(function (Ec2InstanceAction) {
    /**
     * Stop the instance
     */
    Ec2InstanceAction["STOP"] = "stop";
    /**
     * Terminatethe instance
     */
    Ec2InstanceAction["TERMINATE"] = "terminate";
    /**
     * Recover the instance
     */
    Ec2InstanceAction["RECOVER"] = "recover";
    /**
     * Reboot the instance
     */
    Ec2InstanceAction["REBOOT"] = "reboot";
})(Ec2InstanceAction = exports.Ec2InstanceAction || (exports.Ec2InstanceAction = {}));
/**
 * Use an EC2 action as an Alarm action
 */
class Ec2Action {
    constructor(instanceAction) {
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_cloudwatch_actions_Ec2InstanceAction(instanceAction);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, Ec2Action);
            }
            throw error;
        }
        this.ec2Action = instanceAction;
    }
    /**
     * Returns an alarm action configuration to use an EC2 action as an alarm action
     */
    bind(_scope, _alarm) {
        return { alarmActionArn: `arn:${core_1.Stack.of(_scope).partition}:automate:${core_1.Stack.of(_scope).region}:ec2:${this.ec2Action}` };
    }
}
exports.Ec2Action = Ec2Action;
_a = JSII_RTTI_SYMBOL_1;
Ec2Action[_a] = { fqn: "@aws-cdk/aws-cloudwatch-actions.Ec2Action", version: "1.204.0" };
