// @aws-cdk/aws-cloudwatch-actions/lib/ssm.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SsmAction = exports.OpsItemCategory = exports.OpsItemSeverity = void 0;
const jsiiDeprecationWarnings = require("../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("@aws-cdk/core");
/**
 * Types of OpsItem severity available
 */
var OpsItemSeverity;
(function (OpsItemSeverity) {
    /**
     * Set the severity to critical
     */
    OpsItemSeverity["CRITICAL"] = "1";
    /**
     * Set the severity to high
     */
    OpsItemSeverity["HIGH"] = "2";
    /**
     * Set the severity to medium
     */
    OpsItemSeverity["MEDIUM"] = "3";
    /**
     * Set the severity to low
     */
    OpsItemSeverity["LOW"] = "4";
})(OpsItemSeverity = exports.OpsItemSeverity || (exports.OpsItemSeverity = {}));
/**
 * Types of OpsItem category available
 */
var OpsItemCategory;
(function (OpsItemCategory) {
    /**
     * Set the category to availability
     */
    OpsItemCategory["AVAILABILITY"] = "Availability";
    /**
     * Set the category to cost
     */
    OpsItemCategory["COST"] = "Cost";
    /**
     * Set the category to performance
     */
    OpsItemCategory["PERFORMANCE"] = "Performance";
    /**
     * Set the category to recovery
     */
    OpsItemCategory["RECOVERY"] = "Recovery";
    /**
     * Set the category to security
     */
    OpsItemCategory["SECURITY"] = "Security";
})(OpsItemCategory = exports.OpsItemCategory || (exports.OpsItemCategory = {}));
/**
 * Use an SSM OpsItem action as an Alarm action
 */
class SsmAction {
    constructor(severity, category) {
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_cloudwatch_actions_OpsItemSeverity(severity);
            jsiiDeprecationWarnings._aws_cdk_aws_cloudwatch_actions_OpsItemCategory(category);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, SsmAction);
            }
            throw error;
        }
        this.severity = severity;
        this.category = category;
    }
    /**
     * Returns an alarm action configuration to use an SSM OpsItem action as an alarm action
     */
    bind(_scope, _alarm) {
        if (this.category === undefined) {
            return { alarmActionArn: `arn:${core_1.Stack.of(_scope).partition}:ssm:${core_1.Stack.of(_scope).region}:${core_1.Stack.of(_scope).account}:opsitem:${this.severity}` };
        }
        else {
            return { alarmActionArn: `arn:${core_1.Stack.of(_scope).partition}:ssm:${core_1.Stack.of(_scope).region}:${core_1.Stack.of(_scope).account}:opsitem:${this.severity}#CATEGORY=${this.category}` };
        }
    }
}
exports.SsmAction = SsmAction;
_a = JSII_RTTI_SYMBOL_1;
SsmAction[_a] = { fqn: "@aws-cdk/aws-cloudwatch-actions.SsmAction", version: "1.204.0" };
